File: ddss/output.ts
import { Op, type Sequelize } from "sequelize";
import { unparse } from "atsds-bnf";
import { Fact, Idea, initializeDatabase } from "./orm.ts";

export async function main(addr: string, sequelize?: Sequelize) {
    if (!sequelize) {
        sequelize = await initializeDatabase(addr);
    }

    let maxFact = -1;
    let maxIdea = -1;

    while (true) {
        let count = 0;

        const newIdeas = await Idea.findAll({
            where: { id: { [Op.gt]: maxIdea } },
        });
        for (const idea of newIdeas) {
            maxIdea = Math.max(maxIdea, idea.id);
            console.log("idea:", unparse(idea.data));
            count++;
        }

        const newFacts = await Fact.findAll({
            where: { id: { [Op.gt]: maxFact } },
        });
        for (const fact of newFacts) {
            maxFact = Math.max(maxFact, fact.id);
            console.log("fact:", unparse(fact.data));
            count++;
        }

        if (count === 0) {
            await new Promise((resolve) => setTimeout(resolve, 0));
        }
    }
}

File: ddss/output.py
import asyncio
from sqlalchemy import select
from sqlalchemy.ext.asyncio import async_sessionmaker, AsyncSession
from apyds_bnf import unparse
from .orm import Facts, Ideas


async def main(session: async_sessionmaker[AsyncSession]) -> None:
    try:
        max_fact = -1
        max_idea = -1

        while True:
            async with session() as sess:
                for idea in await sess.scalars(select(Ideas).where(Ideas.id > max_idea)):
                    max_idea = max(max_idea, idea.id)
                    print("idea:", unparse(idea.data))
                for fact in await sess.scalars(select(Facts).where(Facts.id > max_fact)):
                    max_fact = max(max_fact, fact.id)
                    print("fact:", unparse(fact.data))
                await sess.commit()

            await asyncio.sleep(0)
    except asyncio.CancelledError:
        pass
